design-documents/graph-ql/mutation-error-design.md

```diff
@@ -0,0 +1,153 @@
+
+## Solution: Design Mutation errors into the schema
+
+If we move our known error states for various mutations into the schema design itself, we'll get a few benefits:
+
+1. Versioning of errors
+2. Discoverability of error states for UI devs
+
+### Example: Fixing our `addItemToCart` mutation
+
+In the previous example of `addItemToCart`, we had the following schema:
+
+```graphql
+type Mutation {
+    addItemToCart(sku: String!, quantity: Float): Cart
+}
+```
+
+Let's design some known error states directly into the API
+
+```graphql
+type Mutation {
+    addItemToCart(sku: String!, quantity: Float): AddItemToCartOutput
+}
+
+type AddItemToCartOutput {
+    # can still get the entire state of the cart post-mutation
+    cart: Cart
+    # can directly access errors that a shopper should be notified about
+    addItemUserErrors: [AddItemUserError!]!
+}
+
+type AddItemUserError {
+    # if a UI has their own text for message, they can just
+    # not use this field, but serves as a descriptive default
+    message: string!
+    type: AddItemUserErrorType!
+}
+
+# This enum is just an example. Non-exhaustive, and of course
+# not all errors we'd want a specific type for
+enum AddItemUserErrorType {
+    OUT_OF_STOCK
+    MAX_QTY_FOR_USER
+    NOT_AVAILABLE
+}
+```
+
+With this new schema, we get the following query and response:
+
+```graphql
+mutation {
+    addItemToCart(sku: "abc", quantity: 1) {
+        cart {
+            items {
+                # select item fields
+            }
+        }
+
+        addItemUserErrors {
+            message
+            type
+        }
+    }
+}
+```
+
+```js
+{
+    "data": {
+        "cart": {
+            // cart data here
+        },
+        "addItemUserErrors": [{
+            "message": "Item 'cool-hat' not added to cart (Out of Stock)",
+            "type": "OUT_OF_STOCK"
+        }]
+    }
+}
+```
+## Should we still use the `errors` key at all?
+
+Yes! _But_, we should consider the `errors` key to be of similar use to the `catch` clause when using `try/catch` in a programming language.
+
+The `errors` key should be for _exceptional_ circumstances, not for describing well-known states in an ecommerce app. It's unlikely you'd write this application code on the backend:
+
+```javascript
+try {
+    var product = addItemToCart('sku');
+    return product;
+} catch (error) {
+    if (error.code === 'OUT_OF_STOCK') {
+        //
+    } else if (error.code === 'MAX_QTY_FOR_USER') {
+        //
+    } else {
+        // something else failed, maybe the DB connection?
+    }
+}
+```
+
+Instead, you would likely have your `addItemToCart` function return something that describes the various possible results of the operation. 
+
+## Open Questions
+
+1. Should we be consistent with the field name that represents these first-class errors? `userErrors` vs something like `addToCartUserErrors`
```

Comment:
I don't think it's necessary to namespace the first class errors.

Reply:
Agree. I might call it `mutationErrors`, but in any case a single global type should suffice.